### PR TITLE
Replace JSON web token with jsrassign, which works in the browser

### PR DESCRIPTION
**Pull request by christiango**
One of the remaining issues with webpack 5 has been with the local server scenarios, where we have the json-web-token dependency. Json-web-token is not meant to run in a browser context, as it is a node library with dependencies like crypto and stream. Webpack 5 no longer automatically polyfills node APIs, so we should move off of json-web-token to an API that works in the browser natively.

This PR proposes to use the jsrsasign library recommended on jwt.io
https://jwt.io/#debugger-io

### CODE REVIEW

**Comment by christiango** [comment]
I'm not sure how to test this locally in the FF repo, one thing I'm noticing is that "@fluidframework/webpack-fluid-loader" does not seem to be pulling the in repo version of "@fluidframework/server-services-client", it seems to be pulling the older version

**Comment by anthony-murphy** [comment]
@fluidframework/server-services-client is in the server mono-repo, and @fluidframework/webpack-fluid-loader is in the client mono-repo. it's possible to link everything into one bit repo with our build tool, --symlink:full see tools\build-tools\README.md

**Comment by christiango** [comment]
> @fluidframework/server-services-client is in the server mono-repo, and @fluidframework/webpack-fluid-loader is in the client mono-repo. it's possible to link everything into one big repo with our build tool, --symlink:full see tools\build-tools\README.md
> 
> But you will need to stage commits across the boundary, and bump the client version after making the server change

Awesome, so should I test this with symlink:full and then do the 2 staged PRs?

**Comment by anthony-murphy** [comment]
yup. exactly

**Review by vladsud — APPROVED** [review]

File: server/routerlicious/packages/services-client/src/auth.ts
Comment (tanviraumi) [inline server/routerlicious/packages/services-client/src/auth.ts:36]: Services also takes a dependency on this generateToken() function. I am assuming using jsrassign in Node.js is okay?

**Review by tanviraumi — APPROVED** [review]

File: server/routerlicious/packages/services-client/src/auth.ts
Comment (christiango) [reply server/routerlicious/packages/services-client/src/auth.ts:36]: It should, how can we test it?

Comment (tanviraumi) [reply server/routerlicious/packages/services-client/src/auth.ts:36]: I did a quick runkit test (https://npm.runkit.com/jsrsasign) and seems to be working with a warning "(node:1) [DEP0005] DeprecationWarning: Buffer() is depreca…, Buffer.allocUnsafe(), or Buffer.from() methods instead".

I was curious about the security aspect when running on service. I personally prefer using well known vetted libraries when it comes to security. But if the jwt.io people recommend this when running on node, I am okay with this too.

Comment (christiango) [reply server/routerlicious/packages/services-client/src/auth.ts:36]: We could keep the service bits unchanged and only update the in browser paths to use the jrassign. Would you prefer that?

json-web-token should never really have been used in a browser context, webpack just happened to bial us out.

File: packages/tools/webpack-fluid-loader/webpack.config.js
Comment (anthony-murphy) [inline packages/tools/webpack-fluid-loader/webpack.config.js:29]: i wonder if this is causing the nconf issue, not sure if that is browser by default, or if it require node

File: server/routerlicious/packages/services-client/src/auth.ts
Comment (anthony-murphy) [reply server/routerlicious/packages/services-client/src/auth.ts:36]: testing should be free, as long as we have mocha test that exercise it, as mocha runs in node :)

File: packages/tools/webpack-fluid-loader/webpack.config.js
Comment (christiango) [reply packages/tools/webpack-fluid-loader/webpack.config.js:29]: It doesn't repro on my laptop, where I never tried to do the sym link full think

Comment (christiango) [reply packages/tools/webpack-fluid-loader/webpack.config.js:29]: Though there is also a "process" implicit dependency, which needs to be sorted out too before merging this that would also block wp5

**Comment by heliocliu** [comment]
Unfortunately we won't get the size analysis until the change makes it to the client repo, but do you happen to know how much the polyfills were contributing to the size/if it's possible to include them explicitly instead of using this other library?  Asking because jsrassign seems on the larger side because it's not using the native browser crypto libs

**Comment by christiango** [comment]
> Unfortunately we won't get the size analysis until the change makes it to the client repo, but do you happen to know how much the polyfills were contributing to the size/if it's possible to include them explicitly instead of using this other library? Asking because jsrassign seems on the larger side because it's not using the native browser crypto libs

I believe this is only for test code, where bundle size is less of a concern (and probably isn't caught by the bundle size test). I don't know how big those other polyfills are, but I do want to make sure whatever change we make here "just works" for consumers of Fluid in webpack 5. What would that change look like?

File: server/routerlicious/packages/services-client/src/auth.ts
Comment (tanviraumi) [reply server/routerlicious/packages/services-client/src/auth.ts:36]: > We could keep the service bits unchanged and only update the in browser paths to use the jrassign. Would you prefer that?

Yeah I would prefer that just to be safe.

> json-web-token should never really have been used in a browser context, webpack just happened to bial us out.

Totally agree.

**Comment by anthony-murphy** [comment]
i also just remembered we added a test to check for isomorphism in the lambdas. This could be a good place ensure your change isn't regressed:
server\routerlicious\packages\lambdas\src\test\webpack.spec.ts

**Comment by christiango** [comment]
> i also just remembered we added a test to check for isomorphism in the lambdas. This could be a good place ensure your change isn't regressed:
> server\routerlicious\packages\lambdas\src\test\webpack.spec.ts

Per Tanvir's comment above, I'm going to not touch the server code with this change.

File: server/routerlicious/packages/services-client/src/auth.ts
Comment (christiango) [reply server/routerlicious/packages/services-client/src/auth.ts:36]: @tanviraumi/ @anthony-murphy  - it looks like tree shaking doesn't work for the json-web-token + server-services-client combo, so we either need to modify server-services-client or break the dependency server-local-server has on server-services-client by perhaps moving GitManager and the name generation modules out into another package. None of these options are great IMO. Thoughts on the best approach here to unblock webpack 5 for consumers of FF?

**Review by vladsud — APPROVED** [review]

File: server/routerlicious/packages/services-client/src/auth.ts
Comment (tanviraumi) [reply server/routerlicious/packages/services-client/src/auth.ts]: Another alternative is to have two different versions of signing token. We move the new jsrassign one to a client specific package and move the jsonwebtoken one to a service specific one (services/services-shared/services-utils). So they will both use their own version. Thoughts?

Comment (christiango) [reply server/routerlicious/packages/services-client/src/auth.ts]: That sounds reasonable, thanks for the suggestion. I'll take a look later today at the server side PR

**Comment by curtisman** [comment]
@heliocliu crypto API is only available in https context. See issue #3869.

**Comment by christiango** [comment]
I had to pull in the new server packages to bring closure to this issue. This also pulled in #4011, which chainged the ITokenClaims contaract. @tanviraumi / @znewton - can you review the parts of this change related to the ITokenClaims and suggest the correct values here?

**Comment by znewton** [comment]
@christiango you beat me to it! #4055 addresses these, but it is still in draft because I did not have the latest server versions available yet. In general, my changes are the same as replacing the spots where you added
```
  iat: 0,
  exp: 0,
```
with
```
  iat: Math.round((new Date()).getTime() / 1000),
  exp: Math.round((new Date()).getTime() / 1000) + 5 * 60, // 5-min expiration window
```

File: packages/tools/webpack-fluid-loader/webpack.config.js
Comment (christiango) [inline packages/tools/webpack-fluid-loader/webpack.config.js:31]: I made webpack-fluid-loader act like webpack 5 so we don't regress this anymore by adding more node dependencies in browser code. The last blocking issue is #4040

**Comment by christiango** [comment]
> @christiango you beat me to it! #4055 addresses these, but it is still in draft because I did not have the latest server versions available yet. In general, my changes are the same as replacing the spots where you added
> 
> ```
>   iat: 0,
>   exp: 0,
> ```
> 
> with
> 
> ```
>   iat: Math.round((new Date()).getTime() / 1000),
>   exp: Math.round((new Date()).getTime() / 1000) + 5 * 60, // 5-min expiration window
> ```

Sweet, I'll update things!

**Comment by christiango** [comment]
Ok, I think this is ready for final review @anthony-murphy / @tanviraumi

File: packages/drivers/routerlicious-urlResolver/src/urlResolver.ts
Comment (christiango) [inline packages/drivers/routerlicious-urlResolver/src/urlResolver.ts:17]: This causes a layer violation. @tanviraumi - is it safe to use jsrsasign here or do we do something else? generateToken is no longer in server-services-client

Comment (christiango) [reply packages/drivers/routerlicious-urlResolver/src/urlResolver.ts:17]: I see there is only 1 usage within the repo - https://github.com/microsoft/FluidFramework/blob/f8cef2c86044f9d2244a9dfe47052e383ca8f525/server/gateway/src/routes/api/api.ts

@jatgarg - I noticed you added this API. We need the client packages to not depend on jsonwebtoken. Is it safe for us to just use generateTokenDirectly from gateway/api.ts rather than depending on a client package for that?

Comment (tanviraumi) [reply packages/drivers/routerlicious-urlResolver/src/urlResolver.ts:17]: I would wait for @jatgarg to confirm my thoughts but this is not the right place to sign a token. Token signing should be done outside the driver. So my recommendation is to make the second parameter in RouterliciousUrlResolver required by removing the "| undefined". Then you can remove the getR11sToken call and also the function. Gateway builds separately so next time when gateway gets updated, people will use the new function in services-utils package. It would be nice if you can put a comment in gateway in this PR.

@jatgarg Thoughts?

File: packages/tools/webpack-fluid-loader/src/loader.ts
Comment (tanviraumi) [inline packages/tools/webpack-fluid-loader/src/loader.ts:36]: I am confused. How does this help the isomorphicity of lambdas? Is it only for local-server?

Comment (christiango) [reply packages/tools/webpack-fluid-loader/src/loader.ts:36]: So, set immediate is not available on modern browsers. When I set the webpack config for webpack-fluid-loader to have the value node: false, this makes it such that webpack-fluid-loader does not include and node polyfills. Doing so would be a build error, which is a good way to make sure we don't regress here.

In terms of why we need this change, this is the last node polyfill we need, so I'm providing it manually until #4040 is resolved. With the move the node: false, coauthoring with local server would fail silently due to the lack of set immediate.

File: examples/utils/get-tinylicious-container/package.json
Comment (christiango) [reply examples/utils/get-tinylicious-container/package.json:34]: Yes, it is in the dependency chain of the all apps in the examples folder

File: packages/drivers/file-driver/src/fileDocumentDeltaConnection.ts
Comment (christiango) [reply packages/drivers/file-driver/src/fileDocumentDeltaConnection.ts:39]: Yup, will fix

File: packages/tools/webpack-fluid-loader/src/loader.ts
Comment (tanviraumi) [reply packages/tools/webpack-fluid-loader/src/loader.ts:36]: Makes sense. So this is not an immediate blocker. Cause we are only allowing one exception, everything else will still be blocked.

**Comment by tanviraumi** [comment]
Should we make the client side token signing a library as well? And may be call it clientAuthGen() or something like that to make it explicit?

**Comment by christiango** [comment]
> Should we make the client side token signing a library as well? And may be call it clientAuthGen() or something like that to make it explicit?

Not sure where something like that would even go. Maybe we should log an issue for that. Sorry, but this PR has already taken up more time than I expected, so I'd like to try to just resolve the getR11sToken and get this in soon

**Comment by tanviraumi** [comment]
> Not sure where something like that would even go. Maybe we should log an issue for that. Sorry, but this PR has already taken up more time than I expected, so I'd like to try to just resolve the getR11sToken and get this in soon

Yeah I am not sure as well. But create an issue may be. @anthony-murphy any idea about a good spot?

File: packages/drivers/routerlicious-urlResolver/src/urlResolver.ts
Comment (christiango) [reply packages/drivers/routerlicious-urlResolver/src/urlResolver.ts]: @jatgarg  - can you take a look, I plan to merge this today

Comment (jatgarg) [reply packages/drivers/routerlicious-urlResolver/src/urlResolver.ts]: Yes @tanviraumi idea looks good to me to make the argument compulsory and not generate token inside the resolver.

File: packages/drivers/local-driver/src/localResolver.ts
Comment (jatgarg) [inline packages/drivers/local-driver/src/localResolver.ts:17]: Is this api going away from this package?

Comment (christiango) [reply packages/drivers/local-driver/src/localResolver.ts:17]: generateToken is already gone from server-services-client. It isn't isomorphic (doesn't work in both browser and server)

**Review by jatgarg — APPROVED** [review]

**Review by tanviraumi — APPROVED** [review]

File: examples/utils/get-tinylicious-container/package.json
Comment (markfields) [reply examples/utils/get-tinylicious-container/package.json:35]: Just to be clear - `@fluidframework/get-tinylicious-container` is a utility package used by the examples, and yes that's part of the client "monorepo".  The Tinylicious package itself is not.

**Pull request merged**
